README.md

```diff
@@ -1,107 +1,149 @@
+entry points.
 
 # <a name="Installation"></a>Installation
 
-BAP is released using `opam` package manager. After you've successfully
+We use OPAM package manager to handle installation. After you've successfully
 [installed](https://opam.ocaml.org/doc/Install.html) opam, do the following:
 
 ```bash
-$ opam init --comp=4.02.1
-$ eval `opam config env`
-$ opam install depext
-$ opam depext bap
-$ opam install bap
+$ opam init --comp=4.02.3    # install the compiler
+$ opam repo add bap git://github.com/BinaryAnalysisPlatform/opam-repository
+$ eval `opam config env`               # activate opam environment
+$ opam depext --install bap            # install bap
 ```
 
+
 # Usage
 
-## Using from OCaml
+## OCaml
 
-There're two ways to use BAP: compile your own stand-alone
-application, and use BAP library, or write a plugin, that can still
-use the library, but will also get an access to decompiled binary, as
-well as intergration with tools and other plugins. For the latter,
-write your plugin in OCaml using your
+An idiomatic way of using BAP is to extend it with a plugin. Suppose,
+you want to write some analysis. For example, let's estimate the ratio
+of jump instructions to the total amount of instructions (a value that
+probably correlates with a complexity of a program).
+
+So, let's do it. Open your
 [favorite text editor](https://github.com/BinaryAnalysisPlatform/bap/wiki/Emacs)
-:
+and write the following program in a file "jmp.ml" (make sure, that
+you created a separate empty directory for your project):
 
-```sh
-$ cat hello_world.ml
+```ocaml
+open Core_kernel.Std
 open Bap.Std
-let main project = print_endline "Hello, World"
-let () = Project.register_pass' "hello-world" main
-```
 
-Next, build it with our `bapbuild` tool:
+let counter = object
+  inherit [int * int] Term.visitor
+  method! enter_term _ _ (jmps,total) = jmps,total+1
+  method! enter_jmp _ (jmps,total) = jmps+1,total
+end
 
-```sh
-$ bapbuild hello_world.plugin
-```
+let main proj =
+  let jmps,total = counter#run (Project.program proj) (0,0) in
+  printf "ratio = %d/%d = %g\n" jmps total (float jmps /. float total)
 
-After this you can load your plugin with `-l` command line option, and
-get an immediate access to the decompiled binary:
 
-```sh
-$ bap /bin/ls -lhello-world
+let () = Project.register_pass' main
 ```
 
-`bapbuild` can compile a standalone applications, not only plugins. In
-fact, `bapbuild` underneath the hood is an `ocamlbuild` utility extended
-with our rules an flags. To compile a standalone binary,
+Before we run it, let's go through the code. The `counter` object is a
+visitor, that has a state consisting of a pair of counters. The first
+counter keeps track of the amount of jmp terms, and the second counter
+is incremented every time we enter any term.  The `main` function, just
+runs the counter. We register it with `Project.register_pass'`, and our
+function will be called by the system, as soon as project is loaded and
```

Comment:
Hmm... actually, after I read the phrase the second time, it looks wrong. It makes people believe, that the pass will be run automatically, albeit that they need to run it explicitly with `--pass`. Will rework it.